ompi_info color coding

gkatev opened the pull request:
This adds colors to ompi_info, per #10813. It's not yet complete but I'm putting it up for any comments thus far.

Todo:
- ~Detect tty for `--color auto`~
- ~Fix the escape characters counting towards the line limit~
- ~Allow customization of colors, probably via MCA param, if possible + try out some other color combinations~

![ompi_info_color_rev0](https://user-images.githubusercontent.com/31203804/191288157-c226163f-4a65-4dc6-91ca-4f63ee0934b4.png)

Fixes #10813
Signed-off-by: George Katevenis <gkatev@ics.forth.gr>

rhc54 commented:
I really like the idea - I only wonder if we might encounter pushback from those users who suffer from color blindness? Perhaps some effort should be made to identify suitable colors that they can also see?

gkatev commented:
Yes of course, if someone has experience and can chime in about more safe/universal colors I can change them up. And if we also implement the customization, the weight of our choice becomes lighter.

rhc54 commented:
I wouldn't say I have experience, but a quick search offered up the following that seem helpful:

https://davidmathlogic.com/colorblind/#%23D81B60-%231E88E5-%23FFC107-%23004D40 had a really good explanation and info on the problem, along with suggestions

https://www.nature.com/articles/nmeth.1618 has a useful chart:

<img width="708" alt="image" src="https://user-images.githubusercontent.com/2440928/191299197-ccfd6074-938d-4d8b-8f72-28a8c349b430.png">

bwbarrett commented on opal/mca/base/mca_base_var.c:
```diff
@@ -2199,10 +2200,22 @@ int mca_base_var_dump(int vari, char ***out, mca_base_var_dump_type_t output_typ
+        char *color_var_valid_values = "", *color_reset = "";
+
+        if(output_type == MCA_BASE_VAR_DUMP_READABLE_COLOR) {
+            color_var_name = "\033[34m";
```
can you add some comments (or better yet, make these defines with descriptive names) on what these colors are?

gkatev replied:
Ok, yes. (this will change up a bit if/when I implement the customization support, so I'll take care of it then)

bwbarrett commented on opal/runtime/opal_info_support.c:
```diff
@@ -197,9 +200,30 @@ int opal_info_init(int argc, char **argv, opal_cmd_line_t *opal_info_cmd_line)
+            opal_cmd_line_get_param(opal_info_cmd_line, "color", 0, 0) : "auto");
+
+        if (0 == strcmp(color, "auto")) {
+            
```
why blank line?  I'm also not sure I love leaving opal_info_color unset here.

gkatev replied:
This is where the ansi/tty check will be implemented, which to-do (the emptyness is misleading indeed, sorry). Generally, auto means colors enabled, unless not supported (no ansi support, pipe etc, not entirely sure how this will be done and if these will be distinct paths)

jsquyres replied:
Check out this:

https://github.com/open-mpi/ompi/blob/ee324be152c1846161dc0553509a0ea4e4534d32/opal/runtime/opal_info_support.c#L873-L889

bwbarrett reviewed (commented):
overall, I like the idea.

jsquyres commented on opal/mca/base/mca_base_var.h:
```diff
@@ -696,10 +696,12 @@ OPAL_DECLSPEC int mca_base_var_build_env(char ***env, int *num_env, bool interna
     /* Dump human-readable strings */
     MCA_BASE_VAR_DUMP_READABLE = 0,
+    /* Dump human-readable strings, with color where supported */
+    MCA_BASE_VAR_DUMP_READABLE_COLOR = 1,
```
Can we add this enum at the end of the list?  I don't *think* that there are going to be any backwards-compatible issues with re-numbering these particular enums, but let's not tempt fate unless we have to.

jsquyres commented on opal/runtime/opal_info_support.c:
```diff
@@ -197,9 +200,30 @@ int opal_info_init(int argc, char **argv, opal_cmd_line_t *opal_info_cmd_line)
+        char *color = (opal_cmd_line_is_taken(opal_info_cmd_line, "color") ?
+            opal_cmd_line_get_param(opal_info_cmd_line, "color", 0, 0) : "auto");
+
+        if (0 == strcmp(color, "auto")) {
```
Might as well make all of these string option comparisons `strcasecmp()`.

gkatev commented:
This site offers a way to visualize what the colors look like under different color blindeness conditions:
https://laura.rochaprado.com/color-blind-simulator/

The param name and param value already seem to be OK. The `valid values` color in my cases above in some cases blends with the blue in param name. Some alternatives below.

valid_values=cyan
<details> 

![vv_cyan](https://user-images.githubusercontent.com/31203804/191752242-c2d931ad-67b4-4188-9288-e96d3d38c70f.png)
</details>

valid_values=red
<details> 

![vv_red](https://user-images.githubusercontent.com/31203804/191752247-0ebea7de-4296-4251-b1c4-f73b0fd4ca4a.png)
</details>

valid_values=yellow
<details>

![vv_yellow](https://user-images.githubusercontent.com/31203804/191752250-a2335a14-efaf-48ee-945f-1d48ca6b7c9a.png)
</details>

I'm open to suggestions about the colors (which ones, and also perhaps what things to color).  Various cases could be made about whether we want the colors to stand out, or how they look under different terminal emulator themes and more.

gkatev commented:
This is supposedly ready for final review.

I wasn't really sure how to name the MCA param, and where to place it. Also not fully sure about the location of the parsing/assisting code, any suggestions welcome.

bosilca commented on opal/mca/base/mca_base_var.c:
```diff
@@ -2060,6 +2084,63 @@ int mca_base_var_get_count(void)
     return mca_base_var_count;
 }
 
+/* Parcer for the opal_var_dump_colors MCA param */
```
parser ?

gkatev replied:
yes indeed... todo

bosilca commented on opal/runtime/opal_params_core.c:
```diff
@@ -211,6 +214,21 @@ int opal_register_util_params(void)
+     * 33=yellow, 34=blue, 35=magenta, 36=cyan, 37=white
+     * https://en.wikipedia.org/wiki/ANSI_escape_code#Colors */
+    opal_var_dump_colors = "name=34:value=32:valid_values=36";
+    ret = mca_base_var_register("opal", "opal", NULL, "var_dump_colors",
```
How would users obtain the list of valid keys ?

gkatev replied:
The default value contains all keys, but indeed, let's write them in the description?

jsquyres commented on opal/mca/base/mca_base_var.c:
```diff
@@ -113,6 +128,11 @@ static const char *info_lvl_strings[] = {"user/basic",  "user/detail",  "user/al
                                          "tuner/basic", "tuner/detail", "tuner/all",
                                          "dev/basic",   "dev/detail",   "dev/all"};
 
+static char *var_dump_color_key_names[VAR_DUMP_COLOR_KEY_COUNT] = {
+    "name", "value", "valid_values"
+};
```
```suggestion
static char *var_dump_color_key_names[VAR_DUMP_COLOR_KEY_COUNT] = {
    [VAR_DUMP_COLOR_VAR_NAME] = "name",
    [VAR_DUMP_COLOR_VAR_VALUE] = "value",
    [VAR_DUMP_COLOR_VALID_VALUES] = "valid_values",
};
```

Using the C99 array initializer syntax guarantees that the strings match the correct index.

gkatev replied:
interesting

gkatev commented:
Addressed the thus-far proposed changes

jsquyres commented:
This is looking pretty good.  I have one question.  Check out this example output from `ompi_info --all`:

![Screen Shot 2022-09-23 at 10 59 19 AM](https://user-images.githubusercontent.com/617316/191991158-a2ae67c7-d9e6-4317-a080-7d6068310edd.png)

* The parameter name is colored: good
* The parameter value is colored: good
* For the enum values, the phrase "Valid values" is colored -- not the permissible values.  Was that intentional?

gkatev commented:
Yes indeed it was intentional, the rationale being something along the lines of not overwhelming the output with too much color. But of course we could change it, all styling suggestions are welcome.

gkatev commented:
Alright yes I can do a quick test tomorrow and show how it would look with the possible values colored, and we can pick whichever seems better.

(I rebased to the latest HEAD)

gkatev commented:
Here is with only the valid values colored:

Dark
<details> 

![dark-v2](https://user-images.githubusercontent.com/31203804/192467442-d5e47f42-767c-4293-afd5-7ffd31cd4cb1.png)

</details> 

Light
<details> 

![light-v2](https://user-images.githubusercontent.com/31203804/192467486-e4d7f8c7-4a1f-4d84-8246-221446dc832e.png)

</details>

I think it looks okay, and beyond the stylistic aspect, we highlight the core points: the parameter, its value, and its possible values (or rather, the values' keys...). This one will entail a small code rearrangement, as it also requires support in mca_base_var_enum.c, but it shouldn't be a problem.

Diff for this test:
```diff
diff --git a/opal/mca/base/mca_base_var.c b/opal/mca/base/mca_base_var.c
index 28658fe55b..204f5883ea 100644
--- a/opal/mca/base/mca_base_var.c
+++ b/opal/mca/base/mca_base_var.c
@@ -2297,7 +2297,8 @@ int mca_base_var_dump(int vari, char ***out, mca_base_var_dump_type_t output_typ
 
             color_var_name = var_dump_colors[VAR_DUMP_COLOR_VAR_NAME];
             color_var_value = var_dump_colors[VAR_DUMP_COLOR_VAR_VALUE];
-            color_valid_values = var_dump_colors[VAR_DUMP_COLOR_VALID_VALUES];
+            // color_valid_values = var_dump_colors[VAR_DUMP_COLOR_VALID_VALUES];
+            color_valid_values = "\033[0m";
             color_reset = "\033[0m";
         }
 
diff --git a/opal/mca/base/mca_base_var_enum.c b/opal/mca/base/mca_base_var_enum.c
index 532b828a5d..914049ee6e 100644
--- a/opal/mca/base/mca_base_var_enum.c
+++ b/opal/mca/base/mca_base_var_enum.c
@@ -421,7 +421,7 @@ static int enum_dump(mca_base_var_enum_t *self, char **out)
 
     tmp = NULL;
     for (i = 0; i < self->enum_value_count && self->enum_values[i].string; ++i) {
-        ret = opal_asprintf(out, "%s%s%d:\"%s\"", tmp ? tmp : "", tmp ? ", " : "",
+        ret = opal_asprintf(out, "%s%s\033[36m%d\033[0m:\"%s\"", tmp ? tmp : "", tmp ? ", " : "",
                             self->enum_values[i].value, self->enum_values[i].string);
         if (tmp) {
             free(tmp);
```

Full image collection: 
[ompi_info_color.zip](https://github.com/open-mpi/ompi/files/9653221/ompi_info_color.zip)

jsquyres commented:
Isn't the string also a valid value for an MCA enum param?  I.e., I thought both the int and the string were valid values.  Meaning: shouldn't both be colored?

I don't quite understand the patch you included in your last comment -- why hard-code `color_valid_values` to `"\033[0m"` instead of `var_dump_colors[VAR_DUMP_COLOR_VALID_VALUES]`?  And in `mca_base_var_enum.c`, why hard-code the colors into the string -- shouldn't those be run-time settable, too?

rhc54 commented:
My impression was that George was just going to do a quick prototype to see how it would look, and then once you decide which you like it, THEN he would do a more permanent coding.

bwbarrett commented:
bot:aws:retest

gkatev commented:
Yes exactly, it is a qucik&dirty change to see which style we prefer, and the diff is there for anyone that also wants to experiment.

Indeed it does sound right to also color the string! Below are examples with int+string colored. The difference of "vva-1" vs "vva-2" is whether the colon between int and string is colored.

vva-1:
<details>

![1](https://user-images.githubusercontent.com/31203804/192720183-d9665abb-1ca0-4ed0-a072-93acc7e1178e.png)
![1-light](https://user-images.githubusercontent.com/31203804/192720399-422e222b-cbb0-4f9c-a2be-83bb772d1fd0.png)
</details>

vva-2:
<details>

![2](https://user-images.githubusercontent.com/31203804/192720474-d626b0a5-01bd-4854-a3fb-3ec31295cd67.png)
![2-light](https://user-images.githubusercontent.com/31203804/192720490-deebc6dc-43f3-4eb3-8568-ce5691c5d10b.png)
</details>

gkatev commented:
I've addeed proper coloring for the actual valid values, but it's not fully there yet. I initially considered using the `mca_base_var_dump_type_t` enum also for the enum dump functions, but seems to be a circular dependency situation going on with the typedefs and the two header files, not fully sure of the best approach yet.

I'm also considering doing the parsing for the colors at an appropriate time instead of on demand when needed, as it bloats the code every where they are needed. I can't do this at `mca_base_var_init()` because the color MCA is not yet available. Would it be "legal" if I attempted to get the value of the `opal_var_dump_colors` MCA right after registering it, or is that too early?

rhc54 commented:
The param value is available immediately after registration, so feel free to use it.

gkatev commented:
This is again ready for review/feedback. I moved the parsing and the color-settings array to opal_params_core. I also renamed the MCA for customizing the color coding to `opal_var_dump_color` (`color`, instead of `colors`).

How it looks now:

<details>

![Untitled](https://user-images.githubusercontent.com/31203804/196897003-a5e6a6d4-ab8b-46a0-8897-f6d2a37029e2.png)
</details>

Useful commands to see all possible coloring scenarios:
```
ompi_info --all --level 9 # general, enum
ompi_info --param opal all --level 9 # bool enum
ompi_info --param btl self --level 9 # flag enum
```

jsquyres reviewed (commented):
I am sorry for taking **forever** to reply to this; I got distracted by some other deadlines. ☹️ 

Thank you for continuing to make progress!  I made a few code suggestions below.  Also, let me *finally* comment on your question in https://github.com/open-mpi/ompi/pull/10820#issuecomment-1260525683: I like `vva-1`, where the colon is not colorized.  I.e., only the actual values are colorized.

I actually take your `vva-1` and raise you: what if we ditch the quotes around string values (since the colors make the quotes kinda moot), and also use `|` between *all* equivalent values.  For example, right now with booleans, we have a somewhat arbitrary `:` after the *first* value, and `|` between all the other equivalent values.  There doesn't seem to be much point in that.  How about something like this:

![Screenshot 2022-11-11 at 12 02 38 PM](https://user-images.githubusercontent.com/617316/201391840-6ab8ede3-27f7-4d54-84e9-4c27df537cb8.png)

![Screenshot 2022-11-11 at 12 03 35 PM](https://user-images.githubusercontent.com/617316/201392015-aacc79e2-f4f4-490b-9c59-5ebfb5c13475.png)

![Screenshot 2022-11-11 at 12 03 58 PM](https://user-images.githubusercontent.com/617316/201392076-51fc0035-a311-4121-9d39-26f2d7d3cb88.png)

If we like this, here's the patch for it.

<details>

```patch
diff --git a/opal/mca/base/mca_base_var_enum.c b/opal/mca/base/mca_base_var_enum.c
index 81553f2b84..dba023f79b 100644
--- a/opal/mca/base/mca_base_var_enum.c
+++ b/opal/mca/base/mca_base_var_enum.c
@@ -123,8 +123,19 @@ static int mca_base_var_enum_bool_dump(mca_base_var_enum_t *self, char **out,
         color_reset = "\033[0m";
     }
 
-    ret = opal_asprintf(out, "%s0: f|false|disabled|no|n%s, %s1: t|true|enabled|yes|y%s",
-                        color_vv, color_reset, color_vv, color_reset);
+    ret = opal_asprintf(out, "%s0%s|%sf%s|%sfalse%s|%sdisabled%s|%sno%s|%sn%s, %s1%s|%st%s|%strue%s|%senabled%s|%syes%s|%sy%s",
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset);
     if (ret < 0) {
         *out = NULL;
         return OPAL_ERR_OUT_OF_RESOURCE;
@@ -228,8 +239,21 @@ static int mca_base_var_enum_auto_bool_dump(mca_base_var_enum_t *self, char **ou
         color_reset = "\033[0m";
     }
 
-    ret = opal_asprintf(out, "%s-1: auto%s, %s0: f|false|disabled|no|n%s, %s1: t|true|enabled|yes|y%s",
-                        color_vv, color_reset, color_vv, color_reset, color_vv, color_reset);
+    ret = opal_asprintf(out, "%s-1%s|%sauto%s, %s0%s|%sf%s|%sfalse%s|%sdisabled%s|%sno%s|%sn%s, %s1%s|%st%s|%strue%s|%senabled%s|%syes%s|%sy%s",
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset,
+                        color_vv, color_reset);
     if (ret < 0) {
         *out = NULL;
         return OPAL_ERR_OUT_OF_RESOURCE;
@@ -335,7 +359,9 @@ static int mca_base_var_enum_verbose_dump(mca_base_var_enum_t *self, char **out,
         color_reset = "\033[0m";
     }
 
-    ret = opal_asprintf(&tmp, "%s, %s0 - 100%s", *out, color_vv, color_reset);
+    ret = opal_asprintf(&tmp, "%s, %s0%s-%s100%s", *out,
+                        color_vv, color_reset,
+                        color_vv, color_reset);
     free(*out);
     if (0 > ret) {
         *out = NULL;
@@ -470,9 +496,10 @@ static int enum_dump(mca_base_var_enum_t *self, char **out,
 
     tmp = NULL;
     for (i = 0; i < self->enum_value_count && self->enum_values[i].string; ++i) {
-        ret = opal_asprintf(out, "%s%s%s%d:\"%s\"%s", tmp ? tmp : "", tmp ? ", " : "",
-                            color_vv, self->enum_values[i].value,
-                            self->enum_values[i].string, color_reset);
+        ret = opal_asprintf(out, "%s%s%s%d%s|%s%s%s",
+                            tmp ? tmp : "", tmp ? ", " : "",
+                            color_vv, self->enum_values[i].value, color_reset,
+                            color_vv, self->enum_values[i].string, color_reset);
         if (tmp) {
             free(tmp);
         }
@@ -768,9 +795,10 @@ static int enum_dump_flag(mca_base_var_enum_t *self, char **out,
     for (int i = 0; i < self->enum_value_count; ++i) {
         tmp = *out;
 
-        ret = opal_asprintf(out, "%s%s%s0x%x:\"%s\"%s", tmp, i ? ", " : " ",
-                            color_vv, flag_enum->enum_flags[i].flag,
-                            flag_enum->enum_flags[i].string, color_reset);
+        ret = opal_asprintf(out, "%s%s%s0x%x%s|%s%s%s",
+                            tmp, i ? ", " : " ",
+                            color_vv, flag_enum->enum_flags[i].flag, color_reset,
+                            color_vv, flag_enum->enum_flags[i].string, color_reset);
         free(tmp);
         if (0 > ret) {
             return OPAL_ERR_OUT_OF_RESOURCE;
```

</details>

gkatev commented on opal/runtime/opal_params_core.c:
```diff
@@ -350,3 +421,71 @@ int opal_register_util_params(void)
+static int parse_color_string(char *color_string, char **key_names,
+        int key_count, char **values_out) {
+
+    for (int i = 0; i < key_count; i++) {
```
Thanks, I like it (wasn't aware of its existence!)

gkatev commented:
Thanks for the feedback, I applied the recommended changes. I think I also lean towards "vva-1", so I applied the patch (it does make the code a little weird with all the printf parameters, but a small price to pay if we like it!). And I also called the raise and applied those changes as well (quotes etc). Theoretically, we could also remove the quotes from the parameter name and value (iff color support is present), but those don't stand out as much; could as well leave them.

jsquyres commented on opal/runtime/opal_params_core.c:
```diff
@@ -350,3 +422,73 @@ int opal_register_util_params(void)
+
+        // Expected format of token: key=code
+        if (opal_argv_count(kv) != 2) {
+            opal_argv_free(kv);
```
Do we need to emit a show_help error here?  I.e., this error case is for when the user has provided a bad string -- we don't want to just silently ignore it, right?

jsquyres commented on opal/runtime/opal_params_core.c:
```diff
@@ -350,3 +422,73 @@ int opal_register_util_params(void)
+
+                break;
+            }
+        }
```
Do we need to emit a show_help error here if kv[0] didn't match anything from key_names?  If not, we'll be silently ignoring a user error.

jsquyres commented on opal/runtime/opal_params_core.c:
```diff
@@ -350,3 +422,73 @@ int opal_register_util_params(void)
+            values_out[k] = strdup(""); // needs to be free-able
+
+            if (NULL == values_out[k]) {
+                for (int v = 0; v < k; v++) {
+                    free(values_out[v]);
+                    values_out[v] = NULL;
+                }
```
I think you can just `opal_argv_free(&values_out)` here...?

Might want to check the overall lifecycle of `opal_var_dump_color` -- both in the error and no-error cases to make sure everything gets freed properly.

gkatev replied:
Looks like it would work due to `values_out[k] == NULL`, but feels hacky since values_out is generally not NULL-terminated.

Generally (in the non-error case) the values are freed in `opal_deregister_util_params` -- I'm not entirely sure at exactly which stage of finalize this gets called. Another alternative could be to free them in `mca_base_var_finalize` (?). In the error case (`OPAL_ERR_OUT_OF_RESOURCE`), theoretically all values are freed, and I expect that the error will propagated and OMPI will fail (and thus `opal_var_dump_color`'s NULL values won't get used in `mca_base_var*`).

jsquyres replied:
> Looks like it would work due to `values_out[k] == NULL`, but feels hacky since values_out is generally not NULL-terminated.

Fair enough.

> Generally (in the non-error case) `opal_var_dump_color` is freed in `opal_deregister_util_params` -- I'm not entirely sure at exactly which stage of finalize this gets called. Another alternative could be to free them in `mca_base_var_finalize` (?). 

That's probably good enough; just as long as they are actually cleaned up properly in the non-error case.

> In the error case (`OPAL_ERR_OUT_OF_RESOURCE`), theoretically all values are freed, and I expect that the error will propagated and OMPI will fail (and thus `opal_var_dump_color`'s NULL values won't get used in `mca_base_var*`).

👍

gkatev commented:
I added help texts for the recommended user-error cases, feel free to suggest alternate forms for them if you have inspiration.

Another small thing we might want to adjust may be the token separator; initially I set it to `:` following the spirit of `LS_COLORS`/`GCC_COLORS` and similar. But perhaps using comma instead might be more intuitive and more in line with other stuff in OMPI?

jsquyres commented on opal/runtime/help-opal-runtime.txt:
```diff
@@ -57,3 +59,11 @@ WARNING: Cannot set both the MCA parameters opal_leave_pinned (a.k.a.,
+#
+[mpi-params:var_dump_color:unknown-key]
+WARNING: Key "%s" in opal_var_dump_color MCA parameter
+token "%s" is unknown, ignoring.
```
Should we also list the valid keys in this error message?  E.g., include something like:

```
Valid keys are:

  %s
```

and the string for the `%s` is built up at run-time and passed in via the call to `opal_show_help()`.

gkatev replied:
Sure, let's. The available valid keys are also in the param description, to which we could point the user when the error occurs, but might as well make it easier and repeat them here too.

Yes, I'd say let's go with `,` for consistency -- indeed the change is trivial. I'll get it done along with the new help text tomorrow.

jsquyres reviewed (commented):
Good question about `:` vs. `,`. Per your point, I think you're right: it might be better to do a comma, since we have several other comma-delimited MCA params.

Thankfully, I think this should be a trivial change in your code (change a single `opal_argv_split()`).

jsquyres commented on opal/runtime/opal_params_core.c:
```diff
@@ -350,3 +422,113 @@ int opal_register_util_params(void)
+            char *valid_keys = NULL, *tmp;
+
+            for (int k = 0; k < key_count; k++) {
+                int ret = opal_asprintf(&tmp, "%s%s\"%s\"", (k > 0 ? valid_keys : ""),
+                    (k > 0 ? ", " : ""), key_names[k]);
+                free(valid_keys);
+                valid_keys = tmp;
```
You could use `opal_argv_join()` to create this string.

gkatev replied:
Didn't know this one either!, I think it's not suitable since key_names is not NULL-terminated, but the `opal_argv_join_range` is usable. I updated the code to use it, in the help text and in the descriptiton of the MCA parameter -- we lose a bit formatting e.g. the space after the comma, but I'd say it's okay.

jsquyres commented:
bot:ompi:retest

jsquyres commented:
bot:aws-v2:retest

jsquyres commented:
@gkatev Thanks for you patience with this PR!

Could you cherry-pick it to v5.0.x?
